README.md
R package sheldus, Losses from Natural Disasters in the United States
========================================================

The SHELDUS database, short for Spatial Hazard Events and Losses Database in the United States [http://webra.cas.sc.edu/hvri/products/sheldus.aspx](http://webra.cas.sc.edu/hvri/products/sheldus.aspx), from the University of South Carolina, is a  database on human and property losses from natural disasters in the United States. Data from this database includes County-level information on property losses, crop losses, injuries and fatalities from 18 different types of natural hazards (hurricanes, droughts, floods, etc.) from about 1960 to the present.

Although the data is free, downloading the data is tedious and so is cleaning and analysing it. The `sheldusr` package comes with the entire data. Users could extract desired data through the functionality provided. Additional graphical and statistical analyses could easily be done in R.

Status of `sheldusr`:
 * Data spans the time period 1960 - 2012. Additional data from SHELDUS 
   (prior to 1960 and Presidential Disaster Declarations) will be 
   included in the future.

 * REMARKS column from the SHELDUS database is ignored for now; will be
   included in the future.

 * `sheldus_extract` can be used to extract data by time period and hazard. 
   Selection by State will be included in the future.
 

NEWS
Version 0.1
-----------

* Data spans the time period 1960 - 2012. Additional data from SHELDUS 
  (prior to 1960 and Presidential Disaster Declarations) will be 
  included in the future.

* `sheldus_extract` can be used to extract data for a desired time period
  and desired hazard. Selection by State will be included in the future.
  
* REMARKS column from the SHELDUS database is ignored for now; will be
  included in the future.
  

R/extract_data.r
#' Extract desired data from pre-processed SHELDUS binary data
#'
#' @param years vector of years for which data is desired, permissible range 1960-2012
#' @param hazards vector of hazards names for which data is desired, permissible hazard names one or more of the following (CASE SENSITIVE!) - "Avalanche", "Coastal", "Drought", "Earthquake", "Flooding", "Fog", "Hail", "Heat", "Hurricane/Tropical Storm", "Landslide", "Lightning", "Severe Storm/Thunder Storm", "Tornado", "Tsunami/Seiche", "Volcano", "Wildfire", "Wind", "Winter Weather"
#' @param inflation logical indicating whether or not inflation adjustment is desired
#' @param base_year year on which the inflation adjustment is based on
#' @export
#' @examples
#' # SHELDUS data for 1996 for all hazards, inflation adjusted for 2010
#' select_hazards <- sheldus_extract(c(1996), inflation = TRUE, base_year = 2010)
#' head(select_hazards)

sheldus_extract <- function(years = c(2012, 2012), hazards = "", 
                            inflation = FALSE, base_year = 2012) {
  
  # global parameters and data
  # years for which SHELDUS database contains data
  year_vec <- c(1960:2012)
  
  # names of hazards from SHELDUS
  hazard_names <- c("Avalanche", "Coastal", "Drought", "Earthquake", "Flooding", 
                    "Fog", "Hail", "Heat", "Hurricane/Tropical Storm", 
                    "Landslide", "Lightning", "Severe Storm/Thunder Storm", 
                    "Tornado", "Tsunami/Seiche", "Volcano", "Wildfire", "Wind", 
                    "Winter Weather")
  
  # invoke data
  sheldus_data <- NULL
  data(sheldus_data, envir = environment())
  
  # sanity checks
  if (sum(nchar(hazards)) > 0) {
    if(!(all(hazards %in% hazard_names))) {
      stop("hazards can be EITHER an empty string OR one or more of - \n", 
           paste(hazard_names, collapse = "\n"), "\n")
    }
  }
  if(!(all(years %in% year_vec))) {
    stop("year has to be within the range ", year_vec[1], " - ", 
         year_vec[length(year_vec)], "\n")
  }
  if (inflation) {
    if(!(base_year %in% sheldus_data$usa_cpi$Year)) {
      stop("year for inflation adjustment should be within the range ", 
           sheldus_data$usa_cpi$Year[1], " - ", sheldus_data$usa_cpi$Year[nrow(sheldus_data$usa_cpi)], "\n")  
    } 
  }

  # extract recrods from the complete data
  out_df <- sheldus_assemble_data(years, hazards, year_vec, hazard_names)
  
  # convert dates to their original format
  out_df$date1 <- as.Date(out_df$date1, origin = "1970-01-01")
  out_df$date2 <- as.Date(out_df$date2, origin = "1970-01-01")
  
  # identify state and county names from fips code
  out_df <- merge(out_df, sheldus_data$usa_fips, by = "fips", all.x = TRUE, sort = FALSE)
  
  # reaname columns to match the original names
  colnames(out_df) <- c("FIPS_CODE", "hazard_id", "HAZARD_BEGIN_DATE", 
                        "HAZARD_END_DATE", "INJURIES", "FATALITIES", 
                        "PROPERTY_DAMAGE", "CROP_DAMAGE", "HAZARD_TYPE_COMBO",
                        "state", "county")
  
  # identify hazard names
  hazard_ids <- sapply(out_df$HAZARD_TYPE_COMBO, 
                       FUN = function (x) grep("1", unlist(strsplit(x, ""))))
  out_df$HAZARD_TYPE_COMBO <- sapply(hazard_ids, 
                                     FUN = function(x) paste(hazard_names[x], 
                                                             collapse = " - "))
  
  # re-order to be in the same order as the original data
  out_df <- out_df[, c("hazard_id", "HAZARD_BEGIN_DATE", "HAZARD_END_DATE",
                       "HAZARD_TYPE_COMBO", "county", "state", "FIPS_CODE", 
                       "INJURIES", "FATALITIES", "PROPERTY_DAMAGE", 
                       "CROP_DAMAGE")]
  
  # adjustment for inflation based on begin year, consitent with SHELDUS
  if (inflation) {
    # identify begin year for each record
    curr_years <- as.POSIXlt(out_df$HAZARD_END_DATE)
    curr_years <- 1900 + curr_years$year    
    
    # cpi for the base year
    base_cpi <- sheldus_data$usa_cpi$CPI[which(sheldus_data$usa_cpi$Year == base_year)]
    # cpi for all the records based on begin year
    curr_cpi <- sapply(curr_years, 
                       FUN = function(x) sheldus_data$usa_cpi$CPI[which(sheldus_data$usa_cpi$Year == x)])
    # adjustment factor
    adj_factor <- base_cpi / curr_cpi
    
    new_prop <- paste0("PROPERTY_DAMAGE_ADJUSTED_", base_year)
    new_crop <- paste0("CROP_DAMAGE_ADJUSTED_", base_year)
    out_df[, new_prop] <- out_df$PROPERTY_DAMAGE * adj_factor
    out_df[, new_crop] <- out_df$CROP_DAMAGE * adj_factor
  }
  
  return (out_df)
}



R/assemble_data.r
#' Read pre-processed SHELDUS data and assemble variables into one data frame
#'
#' @param years vector of years for which data is desired
#' @param hazards vector of hazards names for which data is desired
#' @param year_vec years for which SHELDUS database contains data
#' @param hazard_names names of hazards from SHELDUS
#' @keywords internal

sheldus_assemble_data <- function(years, hazards, year_vec, hazard_names) {
  
  if (sum(nchar(hazards)) > 0) {
    if(!(all(hazards %in% hazard_names))) {
      stop("hazards can be EITHER an empty string OR one or more of of - \n", 
           paste(hazard_names, collapse = "\n"), "\n")
    }
  }
  if(!(all(years %in% year_vec))) {
    stop("year has to be within the range ", year_vec[1], " - ", 
         year_vec[length(year_vec)], "\n")
  }
  
  # invoke data
  sheldus_data <- NULL
  data(sheldus_data, envir = environment())
  
  # process data
  # convert intgert fips to char to 5 digits
  sheldus_data$fips <- sprintf("%.5d", sheldus_data$fips)  
  # combine the 3 sets of hazard data into one 
  data_haz <- paste0(sprintf("%.6d", sheldus_data$hazard1), 
                     sprintf("%.6d", sheldus_data$hazard2), 
                     sprintf("%.6d", sheldus_data$hazard3))
  
  # identify data indices corresp to year
  # convert date of the form "7/1/2011" to int; w.r.t 1970-01-01  
  seq_dates <- c()
  for (eachYear in years) {
    beg_date <- as.numeric(as.Date(paste0("1/1/", eachYear), "%m/%d/%Y"))
    end_date <- as.numeric(as.Date(paste0("12/31/", eachYear), "%m/%d/%Y"))
    seq_dates <- c(seq_dates, seq(beg_date, end_date))
  }
  # dates from the entire dataset present in the desired subset
  index_date <- which(sheldus_data$date2 %in% seq_dates)  
  
  # identify indices of records to be extracted
  if (sum(nchar(hazards)) > 0) {
    # id in the vector of hazard names
    haz_ids <- which(hazard_names %in% hazards)  
    # records from the entire dataset which have the given hazard
    index_haz <- rep(0, length(data_haz))
    for (each_id in haz_ids) {
      index_haz <- index_haz + as.numeric(substr(data_haz, each_id, each_id))  
    }
    index_haz <- which(index_haz > 0)  
    
    index_query <- intersect(index_haz, index_date) 
  } else {
    index_query <- index_date
  }    
  
  # output data frame
  out_df <- data.frame(id = sheldus_data$id[index_query],
                       date1 = sheldus_data$date1[index_query],
                       date2 = sheldus_data$date2[index_query],
                       fips = sheldus_data$fips[index_query],
                       injury = sheldus_data$injury[index_query],
                       fatal = sheldus_data$fatal[index_query],
                       property = sheldus_data$property[index_query],
                       crop = sheldus_data$crop[index_query],
                       hazard = data_haz[index_query],
                       stringsAsFactors = FALSE)
  
  return (out_df)
}


R/sheldusr-package.r
#' sheldusr - Natural disaster losses for the USA from the SHELDUS database
#'
#' The Spatial Hazard Events and Losses Database in the United States 
#' (SHELDUS, http://www.sheldus.org), from the University of South Carolina, 
#' is a database on human and property losses from natural disasters 
#' in the United States. 
#' 
#' Data from this database includes County-level information on property losses, 
#' crop losses, injuries and fatalities from 18 different types of natural 
#' hazards (hurricanes, droughts, floods, etc.) from about 1960 to the present. 
#' 
#' Although the data is free, downloading the data is tedious and so is 
#' cleaning and analysing it. In this package latest data from the SHELDUS 
#' database was packaged compactly for easy access through R.
#' 
#' @references Hazards & Vulnerability Research Institute (2013). 
#' The Spatial Hazard Events and Losses Database for the United States, 
#' Version 12.0 [Online Database]. Columbia, SC: University of South Carolina. 
#' Available from http://www.sheldus.org
#' 
#' @name sheldusr
#' @docType package
NULL


R/data.r
#' SHELDUS data
#'
#' Preprocessed data from SHELDUS and other required information
#' 
#' Fips data was obtained from www.schooldata.com/pdfs/US_FIPS_Codes.xls.
#' 
#' Annual average CPI (all urban consumers) obtained from the US Bureau of 
#' Labor Statistics. See http://www.bls.gov/cpi/tables.htm
#' 
#' HAZARD_TYPE_COMBO could be one or more of 18 hazard types; in the 
#' preprocessed data it is stored as either 1 or 0 depending on the presence or 
#' absence of a hazard; thus, the hazard string for a record is an 18-digit 
#' string of 1s and 0s. During preprocessing this 18-digit string was broken 
#' down into 3 parts of 6 digits each
#'
#' Variables:
#' 
#' \itemize{
#' \item usa_fips - FIPS codes for each county in each state (data frame with 3142 rows and 3 variables)
#'   \itemize{
#'     \item state
#'     \item county
#'     \item fips
#'   }
#'  \item usa_cpi - Consumer Price Index for the United States 1913-2012 (data frame with 100 rows and 2 variables)
#'    \itemize{
#'     \item Year
#'     \item CPI
#'   }
#'  \item id - hazard_id from SHELDUS
#'  \item date1 - HAZARD_BEGIN_DATE from SHELDUS
#'  \item date1 - HAZARD_BEGIN_DATE from SHELDUS
#'  \item fips - FIPS_CODE from SHELDUS
#'  \item injury - INJURIES from SHELDUS
#'  \item fatal - FATALITIES from SHELDUS
#'  \item property - PROPERTY_DAMAGE from SHELDUS
#'  \item crop - CROP_DAMAGE from SHELDUS
#'  \item hazard1 - HAZARD_TYPE_COMBO from SHELDUS
#'  \item hazard2 - HAZARD_TYPE_COMBO from SHELDUS
#'  \item hazard3 - HAZARD_TYPE_COMBO from SHELDUS
#' }
#'
#'
#' @references Hazards & Vulnerability Research Institute (2013). 
#' The Spatial Hazard Events and Losses Database for the United States, 
#' Version 12.0 [Online Database]. Columbia, SC: University of South Carolina. 
#' Available from http://www.sheldus.org
#' 
#' @docType data
#' @name sheldus_data
#' @usage data(sheldus_data)
#' @format List with 13 elements; SHELDUS-preprocessed data each a vector of length 781513
#' @keywords datasets
NULL

